Condition Factory #1

bjoernschroeder opened the pull request:
Adds a conditon factory and replaces the original transition preconditions with generated static functions.

Additional PRs:
Remove Runtime condition objects: https://github.com/rapyuta-robotics/alica/pull/294
Supplementary tests: https://github.com/rapyuta-robotics/alica/pull/292
alica_tests: https://github.com/rapyuta-robotics/alica/pull/291
Turtlesim: https://github.com/rapyuta-robotics/alica/pull/286

athish-t committed:
dummy files

bjoernschroeder committed:
generate condition functions & new creator

bjoernschroeder committed:
add ITransitionPreConditionCreator and TransitionPreConditionFactory

bjoernschroeder committed:
add TransitionPreConditionCreator to AlicaCreators

bjoernschroeder committed:
initialize and provide access to TransitionConditionFactory

bjoernschroeder committed:
run new transition conditions, store inputs of transition conditoins

veerajsk commented on alica_engine/include/engine/ITransitionPreConditionCreator.h:
```diff
@@ -0,0 +1,19 @@
+{
+public:
+    virtual ~ITransitionPreConditionCreator() {}
+    virtual std::function<bool(std::unordered_map<std::string, std::any>)> createConditions(int64_t conditionId) = 0;
```
`const std::unordered_map<std::string, std::any>&`

veerajsk commented on alica_engine/include/engine/ITransitionPreConditionCreator.h:
```diff
@@ -0,0 +1,19 @@
+{
+public:
+    virtual ~ITransitionPreConditionCreator() {}
+    virtual std::function<bool(std::unordered_map<std::string, std::any>)> createConditions(int64_t conditionId) = 0;
```
`const std::unordered_map<std::string, std::any>&`

veerajsk replied:
can typedef this type

veerajsk commented on alica_engine/include/engine/TransitionPreConditionFactory.h:
```diff
@@ -0,0 +1,33 @@
+        auto transitionCondition = _creator.createConditions(id);
+        if (!transitionCondition) {
+            // ALICA_ERROR_MSG("TransitionPreConditionFactory: Condition creation failed: " << id);
+            return nullptr;
```
I assume you are going to change this to a std::function without a target instead

bjoernschroeder replied:
Returns a BasicTransitionCondition now instead of an std::function. However, the factory will be removed.

bjoernschroeder committed:
implement and create TransitionCondition model object

bjoernschroeder committed:
- implement runtime transition condition objects
- store runtime transition condition objects in basicPlans
- create runtime transition objects when the runningPlan is activated
- refactoring

veerajsk commented on alica_tests/src/test/etc/plans/dummy.pml:
```diff
@@ -0,0 +1,100 @@
+      "outState": 1402488477650,
+      "pointsWeb": [],
+	  "condition": "conditions/onSuccess.cond#1532424188199",
+      "conditionRelation": {
```
this should be `keyMapping`. Can you use the latest backend image to generate the pml file?

veerajsk commented on alica_engine/include/engine/ITransitionPreConditionCreator.h:
```diff
@@ -0,0 +1,17 @@
+class RunningPlan;
+class Blackboard;
+
+class ITransitionPreConditionCreator
```
better to name this `ITransitionConditionCreator` since we are dropping the PreCondition name in other places

veerajsk commented on alica_engine/include/engine/BasicTransitionCondition.h:
```diff
@@ -0,0 +1,40 @@
+    std::function<bool(RunningPlan*, Blackboard*)> evalCallback;  
+};
+
+class BasicTransitionCondition : private RunnableObject
```
no need to inherit from `RunnableObject`

veerajsk commented on alica_engine/src/engine/model/Plan.cpp:
```diff
@@ -8,6 +8,7 @@
 #include "engine/model/State.h"
 #include "engine/model/Task.h"
 #include "engine/model/Variable.h"
+#include "engine/BasicTransitionCondition.h"
```
why?

veerajsk reviewed (commented):
Left partial review comments since the PR is still WIP

bjoernschroeder committed:
Merge remote-tracking branch 'origin/rr_devel' into bas_condition_factory

bjoernschroeder committed:
- fix review comments

bjoernschroeder commented:
The RuntimeConditionFactory will be removed and the components of BasicTransitionCondition will be moved to TransitionCondition, the model object. This prevents the creation of objects each time a runningPlan is activated.

bjoernschroeder committed:
- pass WorldModel to transitionCondition evaluation
- move setInput/Output to KeyMapping
- movegetParentWrapperId to RunningPlan

bjoernschroeder committed:
update codegen

veerajsk commented on alica_engine/include/engine/blackboard/KeyMapping.h:
```diff
@@ -18,6 +18,8 @@ class KeyMapping
     const KeyMappingList& getOutputMapping() const;
     void addInputMapping(const std::string& parentKey, const std::string& childKey);
     void addOutputMapping(const std::string& parentKey, const std::string& childKey);
+    void setInput(const Blackboard* parent_bb, Blackboard* child_bb, const KeyMapping* keyMapping) const;
```
no need to take a `keyMapping` parameter. `this` represents the mapping to be done.

veerajsk commented on alica_engine/src/engine/BasicPlan.cpp:
```diff
@@ -73,4 +74,16 @@ int64_t BasicPlan::getId() const
+void BasicPlan::initTransitionConditions(const TransitionGrp& transitions)
+{
+    for (auto it = transitions.begin(); it != transitions.end(); it++) {
+        _transitionConditions[(*it)->getId()] = std::move(_engine->getRuntimeTransitionConditionFactory().create((*it)->getTransitionCondition()));
```
no need of `std::move`. Return values are already prvalues

veerajsk commented on alica_engine/src/engine/RunningPlan.cpp:
```diff
@@ -886,4 +888,30 @@ const KeyMapping* RunningPlan::getKeyMapping(int64_t wrapperId) const
 
+bool RunningPlan::evalTransitionCondition(const Transition* transition, const IAlicaWorldModel* wm)
+{
+    if (isBehaviour()) {
```
should be an assert instead

veerajsk commented on supplementary/alica_codegeneration/alica-plan-designer-fx-codegeneration/src/main/java/de/unikassel/vs/alica/generator/cpp/XtendTemplates.xtend:
```diff
@@ -1121,4 +1121,86 @@ std::shared_ptr<UtilityFunction> UtilityFunction«plan.id»::getUtilityFunction(
+        «IF (protectedRegions.containsKey("condition" + condition.id))»
+«protectedRegions.get("condition" + condition.id)»
+        «ELSE»
+            std::cout << "The condition «condition.id» is not implement yet!" << std::endl;
```
can add a `static_assert(false, "Condition <condition_id> with name <condition_name> is not yet implemented)` so that the error is caught during compilation itself

veerajsk reviewed (commented):
Minor comments. Will approve once the tests are added & working. Can do the optimization of removing RuntimeConditionFactory after tests are added

bjoernschroeder committed:
- Read and store keyMapping of transitions in transition model object
- add DefaultTransitionConditionCreator
- fix getParentWrapperId

bjoernschroeder committed:
fix review comments

Commit:
add name to conditionRepository node

Commit:
move default implementation into default folder

Commit:
- fix includes
- fix creatorCallback names

Commit:
remove todo for constraints

Commit:
update codegeneration

Commit:
use correct TransitionConditions for generating callbacks, remove transition generation from plans

Commit:
remove tmp node from modelmanager

Commit:
update TestContext to use TransitionConditionCreator

Commit:
remove old code, add protected region to conditions header, update codegen jar

bjoernschroeder committed:
- remove BasicTransitionCondition (#294)

- move blackboard and callback to TransitionCondition model object
- replace RuntimeTransitionConditionFactory with TransitionConditionCallbackFactory
- set evalCallback for TransitionConditions in engine::init()
- remove BasicTransitionCondition from BasicPlans

Co-authored-by: bjoernschroeder <bjoernschroder@rapyuta-robotics.com>

abhi1287 commented on alica_engine/include/engine/AlicaEngine.h:
```diff
@@ -87,7 +88,9 @@ class AlicaEngine
     Blackboard& editBlackboard() { return _Blackboard; }
-
+  
+    const TransitionConditionCallbackFactory& getTransitionConditionCallbackFactory() const { return *_transitionConditionCallbackFactory; }
```
tocheck: unused?

abhi1287 commented on alica_engine/src/engine/AlicaEngine.cpp:
```diff
@@ -123,6 +128,14 @@ void AlicaEngine::terminate()
+{
+    for (const Transition* transition : _planRepository.getTransitions()) {
+        TransitionCondition* transitionCondition = transition->getTransitionCondition();
+        transitionCondition->setEvalCallback(_transitionConditionCallbackFactory->create(transitionCondition));
```
circular dependency, just pass what the method needs i.e. name and id

abhi1287 commented on alica_engine/include/engine/TransitionConditionCallbackFactory.h:
```diff
@@ -0,0 +1,28 @@
+class TransitionCondition;
+class BasicTransitionCondition;
+
+class TransitionConditionCallbackFactory
```
Why do we need this? Can't we call creator directly and pass necessary arguments to the creator create() function

abhi1287 commented on alica_engine/include/engine/blackboard/KeyMapping.h:
```diff
@@ -18,6 +18,8 @@ class KeyMapping
     void addInputMapping(const std::string& parentKey, const std::string& childKey);
     void addOutputMapping(const std::string& parentKey, const std::string& childKey);
+    void setInput(const Blackboard* parent_bb, Blackboard* child_bb) const;
+    void setOutput(Blackboard* parent_bb, const Blackboard* child_bb) const;
```
seems out of place, key mapping shouldn't know about blackboard. its just a mapping

abhi1287 commented on alica_engine/include/engine/model/TransitionCondition.h:
```diff
@@ -0,0 +1,40 @@
+struct TransitionConditionContext
+{
+    std::unique_ptr<BlackboardBlueprint> blackboardBlueprint;
+};
```
why to wrap this in struct? can be directly passed to TransitionCondition.

abhi1287 commented on alica_engine/src/engine/model/TransitionCondition.cpp:
```diff
@@ -0,0 +1,25 @@
+        return false;
+    }
+    keyMapping->setInput(rp->getBasicPlan()->getBlackboard().get(), _blackboard.get());
+    return _evalCallback(_blackboard.get(), rp, wm);
```
assert that the callback exists.

abhi1287 reviewed (commented):
please fix the names, there are 10+ names with TransitionCondition* prefix. Then TransitionConditionCallbackFactory is calling TransitionConditionCreator to create TransitionConditionCallback. 
I feel many of them could be skipped altogether

Commit:
- remove TransitionConditionCallbackFactory
- create TransitionCondition in initTransitionConditions in engine
- remove TransitionConditionContext struct
- assert that TransitionCondition callback exists before calling

Commit:
remove forward declaration

Commit:
move setInput / setOutput to BlackboardUtil

Commit:
add iostream and debug_output includes

Commit:
Merge remote-tracking branch 'origin/rr_devel' into bas_condition_factory

abhi1287 committed:
Merge branch 'rr_devel' into bas_condition_factory

Commit:
Merge remote-tracking branch 'origin/rr_devel' into bas_condition_factory

bjoernschroeder committed:
Condition Factory #3 Update turtlesim for new conditions (#286)

* - regenerate code
- move implementation of transitions to conditions.cpp
- update base

* regenerate files

* fix turtlesim after merge, move addSolver call after initialization of context

* Condition Factory #4 Update alica_tests to use new conditions (#291)

* update tests to use new conditions

* update TestContext

* regenerate

* Condition Factory #5 Update supplementary tests for new transition conditions (#292)

* use new TransitionConditons

* regenerate code, move generated files from autogenerated to Expr

* format

Co-authored-by: bjoernschroeder <bjoernschroder@rapyuta-robotics.com>

Co-authored-by: bjoernschroeder <bjoernschroder@rapyuta-robotics.com>

Co-authored-by: bjoernschroeder <bjoernschroder@rapyuta-robotics.com>

triccyx commented on alica_engine/include/engine/blackboard/BlackboardUtil.h:
```diff
@@ -0,0 +1,12 @@
+    static void setInput(const Blackboard* parent_bb, Blackboard* child_bb, const KeyMapping* keyMapping);
+    static void setOutput(Blackboard* parent_bb, const Blackboard* child_bb, const KeyMapping* keyMapping);
+};
+} // namespace alica
```
Minor: github ask for an empty line at the end of every files

triccyx commented on alica_engine/include/engine/default/DefaultTransitionConditionCreator.h:
```diff
@@ -0,0 +1,19 @@
+public:
+    DefaultTransitionConditionCreator() = default;
+    ~DefaultTransitionConditionCreator() = default;
+    TransitionConditionCallback createConditions(std::string name) const;
```
You can use:
`const std::string & `
here to avoid a copy

triccyx commented on alica_engine/include/engine/default/DefaultTransitionConditionCreator.h:
```diff
@@ -0,0 +1,19 @@
+    DefaultTransitionConditionCreator() = default;
+    ~DefaultTransitionConditionCreator() = default;
+    TransitionConditionCallback createConditions(std::string name) const;
+    bool isDefaultTransitionCondition(std::string name) const;
```
You can use:
`const std::string & `
here to avoid a copy

triccyx commented on alica_engine/include/engine/default/DefaultTransitionConditionCreator.h:
```diff
@@ -0,0 +1,19 @@
+class DefaultTransitionConditionCreator
+{
+public:
+    DefaultTransitionConditionCreator() = default;
```
Can be removed if empty the compiler will create as default

triccyx commented on alica_engine/include/engine/default/DefaultTransitionConditionCreator.h:
```diff
@@ -0,0 +1,19 @@
+{
+public:
+    DefaultTransitionConditionCreator() = default;
+    ~DefaultTransitionConditionCreator() = default;
```
Can be removed if empty the compiler will create as default

triccyx commented on alica_engine/include/engine/model/TransitionCondition.h:
```diff
@@ -0,0 +1,36 @@
+{
+public:
+    TransitionCondition(std::unique_ptr<BlackboardBlueprint> blackboardBlueprint);
+    virtual ~TransitionCondition() = default;
```
Can be removed if you don't plan to subclass

triccyx commented on alica_engine/include/engine/model/TransitionCondition.h:
```diff
@@ -0,0 +1,36 @@
+public:
+    TransitionCondition(std::unique_ptr<BlackboardBlueprint> blackboardBlueprint);
+    virtual ~TransitionCondition() = default;
+    virtual bool evaluate(const RunningPlan* rp, const IAlicaWorldModel* wm, const KeyMapping* keyMapping);
```
virtual can be removed if you don't plan to subclass

triccyx commented on alica_engine/include/engine/model/TransitionCondition.h:
```diff
@@ -0,0 +1,36 @@
+
+private:
+    // TODO: Check if friends are necessary
+    friend ModelFactory;
```
Can be removed

triccyx commented on alica_engine/include/engine/model/TransitionCondition.h:
```diff
@@ -0,0 +1,36 @@
+private:
+    // TODO: Check if friends are necessary
+    friend ModelFactory;
+    friend TransitionConditionFactory;
```
Can be removed

triccyx commented on alica_engine/include/engine/model/TransitionConditionRepository.h:
```diff
@@ -0,0 +1,30 @@
+    std::string getFileName() const;
+
+private:
+    friend ModelFactory;
```
It seems that these friends can be removed

triccyx commented on alica_engine/include/engine/model/TransitionConditionRepository.h:
```diff
@@ -0,0 +1,30 @@
+    TransitionConditionRepository();
+    virtual ~TransitionConditionRepository();
+    const TransitionConditionGrp& getTransitionConditions() const { return _transitionConditions; }
+    std::string getFileName() const;
```
Evalutate to substitute return with 
`const std::string &`

triccyx commented on alica_engine/src/engine/blackboard/BlackboardUtil.cpp:
```diff
@@ -0,0 +1,38 @@
+
+namespace alica
+{
+void BlackboardUtil::setInput(const Blackboard* parent_bb, Blackboard* child_bb, const KeyMapping* keyMapping)
```
As they are shared_ptr I would consider passing them instead of the .get() also as references. And if necessary do .get() inside this method.

triccyx commented on alica_engine/src/engine/blackboard/BlackboardUtil.cpp:
```diff
@@ -0,0 +1,38 @@
+    }
+}
+
+void BlackboardUtil::setOutput(Blackboard* parent_bb, const Blackboard* child_bb, const KeyMapping* keyMapping)
```
As they are shared_ptr I would consider passing them instead of the .get() also as references.  And if necessary do .get() inside this method.

triccyx commented on alica_engine/src/engine/default/DefaultConditions.cpp:
```diff
@@ -0,0 +1,13 @@
+
+namespace alica
+{
+bool defaultCondition(const Blackboard* input, const RunningPlan* rp, const IAlicaWorldModel* wm)
```
As they are shared_ptr I would consider passing them instead of the .get(). And if necessary do .get() inside this method.

bjoernschroeder replied:
I talked with veeraj about passing shared_ptr instead of raw ptrs and we decided to stick to the raw ptrs for internally called functions

triccyx commented on alica_engine/src/engine/default/DefaultTransitionConditionCreator.cpp:
```diff
@@ -0,0 +1,48 @@
+namespace alica
+{
+
+TransitionConditionCallback DefaultTransitionConditionCreator::createConditions(std::string name) const
```
const std::string& to avoid copy

triccyx commented on alica_engine/src/engine/default/DefaultTransitionConditionCreator.cpp:
```diff
@@ -0,0 +1,48 @@
+    throw new std::exception();
+}
+
+bool DefaultTransitionConditionCreator::isDefaultTransitionCondition(std::string name) const
```
const std::string& to avoid copy

triccyx commented on alica_engine/src/engine/modelmanagement/factories/TransitionConditionFactory.cpp:
```diff
@@ -0,0 +1,26 @@
+
+namespace alica
+{
+TransitionCondition* TransitionConditionFactory::create(const YAML::Node& conditionNode, TransitionConditionRepository* conditionRepository)
```
conditionRepository not used can be removed

bjoernschroeder replied:
We need the conditionRepository because ModelManager::parseFile requires us to return a single AlicaElement. Because all conditions are stored in a single file and we cant return a vector of conditions, we store the conditions in the conditionsRepository.

triccyx commented on alica_engine/src/engine/model/TransitionConditionRepository.cpp:
```diff
@@ -0,0 +1,20 @@
+
+TransitionConditionRepository::~TransitionConditionRepository() {}
+
+std::string TransitionConditionRepository::getFileName() const
```
Consider using const std::string&

triccyx commented on alica_engine/src/engine/modelmanagement/factories/TransitionFactory.cpp:
```diff
@@ -54,5 +59,12 @@ void TransitionFactory::attachReferences()
     }
     Factory::transitionSynchReferences.clear();
+    // transitionConditionReferences
+    for (std::pair<int64_t, int64_t> pairs : Factory::transitionConditionReferences) {
```
consider using reference here:
std::pair<int64_t, int64_t>& pairs

triccyx reviewed (changes requested):
For me some minor comments.
Do not hesitate to comment on them if you do not agree.

Commit:
fix review comments

Commit:
format

triccyx reviewed (approved):
Ok for me
